github-email-notifications/githubconsumer.py

```diff
@@ -1,40 +1,148 @@
+        print(
+            "{comment_author} commented on a pull request".format(**comment),
+            file=output
+        )
+        print(comment['comment_body'], file=output)
+        print(
+            "See the full comment at {comment_url}\n".format(**comment),
+            file=output
+        )
+        res = output.getvalue()
+        output.close()
+        return res
+
+    @abstractmethod
     def fmt_pr(self, pull_req):
-        # this could be mail subject
-        print "{pr_author}'s pull request #{pr_num}: \"{pr_title}\" was {pr_action}\n".format(**pull_req)
-        # this could be mail body
+        output = StringIO.StringIO()
+        print(
+            "{pr_author}'s pull request #{pr_num}: \"{pr_title}\" was "
+            "{pr_action}\n".format(**pull_req),
+            file=output
+        )
         if pull_req['pr_action'] == u'opened':
-            print "PR body:\n{pr_body}\n".format(**pull_req)
-        print "See the full pull-request at {pr_url}\n".format(**pull_req)
+            print("PR body:\n{pr_body}\n".format(**pull_req), file=output)
+        print(
+            "See the full pull-request at {pr_url}\n".format(**pull_req),
+            file=output
+        )
+        res = output.getvalue()
+        output.close()
+        return res
 
-class RawPPFormatter(object):
+
+class StdoutFormatter(Formatter):
+    def fmt_issue_comment(self, comment):
+        print(super(StdoutFormatter, self).fmt_issue_comment(comment))
+
+    def fmt_pr(self, pull_req):
+        print(super(StdoutFormatter, self).fmt_pr(pull_req))
+
+
+class EmailFormatter(Formatter):
+    def __init__(self, to_addr, from_addr, smtp_server):
+        """
+
+        :param to_addr: email destinations
+        :param from_addr: source email address
+        """
+        super(EmailFormatter, self).__init__()
+
+        self.to_addr = to_addr
+        self.from_addr = from_addr
+        self.smtp_server = smtp_server
+
+    def _send_email(self, subject, body, attachment=None, new=False):
+        """
+        Inspired by: https://github.com/abartlet/gh-mailinglist-notifications/blob/master/gh-mailinglist.py
+        :param subject:
+        :param body:
+        :param attachment:
+        :param new:
+        :return:
+        """
+        subject = subject.replace('\n', ' ').replace('\r', ' ')
+        outer = MIMEMultipart()
+        outer['Subject'] = Header(subject, 'utf8')
+        outer['To'] = self.to_addr
+        outer['From'] = self.from_addr
+
+        # FIXME messageid for threads
+        #        msgid = "<gh-mailinglist-notifications-%s-%s-%d@%s>" % (
+        #            uuid, repo, pr_number, msg_id_domain)
+        #        if new:
+        #            outer.add_header("Message-ID", msgid)
+        #        else:
+        #            outer.add_header("In-Reply-To", msgid)
+        #            outer.add_header("References", msgid)
+
+        # FIXME attachment (patches, CI output?)
```

Comment:
OK, whatever works for me, it is easy to attach patch and even easier to
ignore patch in email. But I still think that only one needed archeology is
git itself, I never searched for older versions of pushed patches

Dňa 12. 8. 2016 5:19 odpoludnia používateľ "Petr Špaček" <
notifications@github.com> napísal:

> In github-email-notifications/githubconsumer.py
> <https://github.com/freeipa/freeipa-tools/pull/4#discussion_r74607642>:
>
> > +        subject = subject.replace('\n', ' ').replace('\r', ' ')
> > +        outer = MIMEMultipart()
> > +        outer['Subject'] = Header(subject, 'utf8')
> > +        outer['To'] = self.to_addr
> > +        outer['From'] = self.from_addr
> > +
> > +        # FIXME messageid for threads
> > +        #        msgid = "<gh-mailinglist-notifications-%s-%s-%d@%s>" % (
> > +        #            uuid, repo, pr_number, msg_id_domain)
> > +        #        if new:
> > +        #            outer.add_header("Message-ID", msgid)
> > +        #        else:
> > +        #            outer.add_header("In-Reply-To", msgid)
> > +        #            outer.add_header("References", msgid)
> > +
> > +        # FIXME attachment (patches, CI output?)
>
> I was thinking that it is handy to have patches attached for
> archaeological purposes (to have e-mails together with patches in mailing
> list archive). Petr^2 Spacek
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/freeipa/freeipa-tools/pull/4/files/2b954463d9ed625ed2574c6bc1a73f31a29429ad#r74607642>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AECrYfj4PrRoTt3reD5YqcZojiOY8mrJks5qfI1ogaJpZM4JjB63>
> .
>